homeassistant/components/tuya/__init__.py

```diff
@@ -257,7 +257,11 @@ def update_device(self, device: TuyaDevice) -> None:
                 device.id,
                 self.device_manager.device_map[device.id].status,
             )
-            dispatcher_send(self.hass, f"{TUYA_HA_SIGNAL_UPDATE_ENTITY}_{device.id}")
+
+            if not self.hass.loop.is_closed():
```

Comment:
Do we need to remove the device listener on Home Assistant stop event? I don't see that being done anywhere in the current code. We do it when unloading the entry but not when stopping Home Assistant.

Reply:
The reason for adding the `if` was because it's being called wheb HA is being restarted (process of shutdown), as result of loop being terminated it throws an ex exception as described in the bug related to the PR

Reply:
Yes, and that suggest that we should stop and remove the device listener when stopping Home Assistant, instead of adding the check here. What do you think?

Reply:
Agree, that the listener should be removed when stopping.

Reply:
not sure I follow, this function is being called as part of an update_device operation,
it can take place (and happens) during restart of HA - code should have protection that if the hass.loop is closed, with that approach dispatcher will not get called on restart even if got an update request the same millisecond

on unload of component the device listener is being removed:
```python
hass_data.device_manager.remove_device_listener(hass_data.device_listener)
```

what should we do with it?

Reply:
We should do the same clean up of this listener as during config entry unload but for the `EVENT_HOMEASSISTANT_STOP`.

Register a listener that fires once for this event and does the clean up. Search for `EVENT_HOMEASSISTANT_STOP` to see examples.

Reply:
@elad-bar here is how we do it for Shelly:
`async_on_unload` makes sure we don't call shutdown in case of unload since unload entry also calls shutdown
https://github.com/home-assistant/core/blob/1d6e40451274cbe8ccc6be256146193bcfe97df5/homeassistant/components/shelly/__init__.py#L316-L317
https://github.com/home-assistant/core/blob/1d6e40451274cbe8ccc6be256146193bcfe97df5/homeassistant/components/shelly/__init__.py#L495-L497